inference-engine/src/cldnn_engine/cldnn_infer_request.cpp

```diff
@@ -958,77 +925,125 @@ void copyToFloat(float* dst, const InferenceEngine::Blob* src) {
+    }
+    Blob::Ptr reqBlob = _gpuInputs.at(inputName);
     auto _nw_ptr = m_graph->GetNetwork();
-    auto prec = inputBlob.getTensorDesc().getPrecision();
-
-    if (inputBlob.is<gpu::ClBlob>()) {
-        // no need to check for reuse
-        _nw_ptr->set_input_data(internalName, memory);
-    } else if (prec == Precision::I16 || prec == Precision::U16) {
-        // clDNN doesn't support I16 input precision, so we always have to convert input data to fp32 precision
-        const cldnn::memory& fp32_mem = inputsMemory.at(inputName+fp32_suffix);
-        cldnn::pointer<float> ptr = fp32_mem.pointer<float>();
-        if (prec == Precision::I16) {
-            copyToFloat<int16_t>(ptr.data(), &inputBlob);
-        } else {
-            copyToFloat<uint16_t>(ptr.data(), &inputBlob);
+    cldnn::primitive_id internalName = "parameter:" + inputName;
+    const auto& prec = inputBlob->getTensorDesc().getPrecision();
+    auto remote_ptr = inputBlob->as<gpu::ClBlob>();
+    if (remote_ptr != nullptr) {
+        switch (prec) {
+        case Precision::FP32:
+        case Precision::FP16:
+        case Precision::I8:
+        case Precision::U8:
+        case Precision::BOOL:
+        case Precision::I32:
+        case Precision::I64: {
+            auto impl = getBlobImpl(remote_ptr);
+            if (!impl->is_allocated()) {
+                IE_THROW() << str_input_not_allocated;
+            }
+            auto inputMem = impl->getMemory();
+            _nw_ptr->set_input_data(internalName, inputMem);
+            break;
+        }
+        default:
+            IE_THROW() << "Unsupported input precision " << prec;
+        }
+    } else {
+        auto impl = getBlobImpl(reqBlob->as<gpu::ClBlob>());
+        if (!impl->is_allocated()) {
+            IE_THROW() << str_input_not_allocated;
         }
+        auto inputMem = impl->getMemory();
 
-        _nw_ptr->set_input_data(internalName, fp32_mem);
-    } else if (is_same_buffer(inputBlob, memory)) {
-        // If input memory was allocated by cldnn engine and wasn't overwritten by user set_input_data method won't copy input data.
-        switch (prec) {
-            case Precision::FP32:
-            case Precision::FP16:
-            case Precision::I8:
-            case Precision::U8:
-            case Precision::BOOL:
-            case Precision::I32:
+        if (prec == Precision::I16 || prec == Precision::U16) {
+            // clDNN doesn't support I16 input precision, so we always have to convert input data to fp32 precision
+            cldnn::pointer<float> ptr = inputMem.pointer<float>();
+            if (prec == Precision::I16) {
+                copyToFloat<int16_t>(ptr.data(), inputBlob.get());
+            } else {
+                copyToFloat<uint16_t>(ptr.data(), inputBlob.get());
+            }
+        } else {
+            switch (prec) {
+            case Precision::FP32: {
+                copyToMem<float, float>(inputMem, inputBlob.get());
```

Comment:
Why do you need typed copy here? Won't simple bites copy work?

Reply:
Fixed